# Welcome to the GEDI

Gedi is a software platform for handling genomic data such as sequencing reads, sequences, per-base numeric values or annotations written in Java.

Gedi is developed in the group of Florian Erhard at the [Institute for Virology and Immunobiology](http://www.virologie.uni-wuerzburg.de/ags_virologie/ag_erhard/) at the University of WÃ¼rzburg.

Its main feature are:
* Comprehensive general purpose software library (Managing iterators, Serialization, Random access I/O, Parallelization, Extension system, JS based template engine, JSON, Strings, Arrays, ...)
* Specialized algorithms and data structures for Bioinformatics (Clustering, maximum scoring subsequences, sequence alignment, string searching, suffix trees, tries, union find, range minimum queries, ...) and Statistics (Descriptive statistics, inference, kernel methods, regression)
* Even more specialized algorithms and data structures for genomic data (Random accessed fasta files, Memory based interval trees, disk based interval trees, space efficient handling of aligned reads, Annotation management, Id mapping,...)
* Graphical user interface (Project management, Table view, Genome browser, ...)

Most of the tools and methods developed in the group are part of, or a plugin to Gedi (e.g. [LFC](https://www.ncbi.nlm.nih.gov/pubmed/26160885), PRICE).
